src/Magnum/Octree/Octree.h

```diff
@@ -0,0 +1,274 @@
+ */
+
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Range.h>
+#include <Magnum/Math/Vector3.h>
+#include <Magnum/Math/Vector4.h>
+#include <Magnum/Shapes/AxisAlignedBox.h>
+
+#include <vector>
+#include "visibility.h"
+
+namespace Magnum { namespace Math { namespace Geometry { namespace Intersection {
+
+/**
+ * @brief Which half space of the plane the point is in.
+ * @param point the point
+ * @param plane the plane
+ * @returns a value `> 0` if on positive half space,
+ *                  `< 0` if on negative half space,
+ *                  `0` if on the plane.
+ */
+template<class T>
+float pointPlaneHalfSpace(Vector3<T> point, Vector4<T> plane) {
+    return (plane.xyz()*point).sum() + plane.w();
+}
+
+template<class T>
+bool pointFrustum(Vector3<T> point, const Vector4<T> (&frustum)[6]) {
+    for(const Vector4<T>& f : frustum) {
+        if(pointPlaneHalfSpace(point, f) < 0) {
+            /* the point is in front of one of the frustum planes (normals point outwards) */
+            return false;
+        }
+    }
+
+    return true;
+}
+
+template<class T>
+bool rangeFrustum(const Range3D<T>& box, const Vector4<T> (&frustum)[6]) {
+    /* Create the 8 vertices of the box from the 2 given vertices min and max */
+    /* Check for each corner of an octant whether it is inside the frustum.
+     * If only some of the corners are inside, the octant requires further checks. */
+    int planes = 0;
+
+    for(const Vector4<T>& plane : frustum) {
+        int corners = 0;
+
+        for(UnsignedByte c = 0; c < 8; ++c) {
+            const Vector3<T> corner = Math::lerp(box.min(), box.max(), Math::BoolVector<3>{c});
+
+            if(Math::Geometry::Intersection::pointPlaneHalfSpace<Float>(corner, plane) >= 0) {
+                ++corners;
+            }
+        }
+
+        if(corners == 0) {
+            /* all corners are outside this plane */
+            return false;
+        }
+
+        if(corners == 8) {
+            ++planes;
+        }
+    }
+
+    if(planes == 6) {
+        return true;
+    }
+
+    /* potentially check corners here! */
+    return true;
+}
+
+}}}}
+
+namespace Magnum { namespace Octree {
+
+
+enum class OctantStatus: Byte {
+    NoCorners = 0,   /**< No corners of the octant are inside the testing volume */
+    SomeCorners = 1, /**< Neither all nor no corners of the octant are inside the testing volume */
+    AllCorners = 2,  /**< All corners of the octant are inside the testing volume */
+};
+
+/**
+ * @brief Octree @see https://en.wikipedia.org/wiki/Octree
+ *
+ * An octree is a cubic structure, which is subdivided in up to 8 smaller cubes,
+ * called octants. Each octant can be seen as an octree itself. Data is stored
+ * in the leaf octants (octants without children).
+ */
+template<typename T>
+class Octree {
+    public:
+
+        /** @brief Default constructor */
+        Octree() :
+            _center{},
+            _radius(),
+            _depth(),
+            _isLeafNode()

```

Comment:
I don't care about old compilers anymore and I think in this case doing in-class initialization of the members would be better than here in constructor.